pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -472,20 +472,93 @@ class StateVectorKokkos final
                         "`controlled_wires` must have the same size as "
                         "`controlled_values`.");
-        PL_ABORT_IF_NOT(
-            array_contains(controlled_gate_names, std::string_view{opName}),
-            "Controlled matrix operation not yet supported.");
 
         if (controlled_wires.empty()) {
             return applyOperation(opName, wires, inverse, params, gate_matrix);
         }
+        if (array_contains(controlled_gate_names, std::string_view{opName})) {
+            const std::size_t num_qubits = this->getNumQubits();
+            const ControlledGateOperation gateop =
+                reverse_lookup(controlled_gate_names, std::string_view{opName});
+            applyNCNamedOperation<KokkosExecSpace>(
+                gateop, *data_, num_qubits, controlled_wires, controlled_values,
+                wires, inverse, params);
+        } else {
+            PL_ABORT_IF(gate_matrix.empty(),
+                        std::string("Operation does not exist for ") + opName +
+                            std::string(" and no matrix provided."));
+            return applyNCMultiQubitOp(vector2view(gate_matrix),
+                                       controlled_wires, controlled_values,
+                                       wires, inverse);
+        }
+    }
 
-        const std::size_t num_qubits = this->getNumQubits();
-        const ControlledGateOperation gateop =
-            reverse_lookup(controlled_gate_names, std::string_view{opName});
-        applyNCNamedOperation<KokkosExecSpace>(
-            gateop, *data_, num_qubits, controlled_wires, controlled_values,
-            wires, inverse, params);
+    /**
+     * @brief Apply a controlled-multi qubit operator to the state vector using
+     * a matrix
+     *
+     * @param matrix Kokkos gate matrix in the device space.
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (true or false).
+     * @param wires Wires to apply gate to.
+     * @param inverse Indicates whether to use adjoint of gate. (Default to
+     * false)
+     */
+    void applyNCMultiQubitOp(const KokkosVector matrix,
+                             const std::vector<std::size_t> &controlled_wires,
+                             const std::vector<bool> &controlled_values,
+                             const std::vector<std::size_t> &wires,
+                             bool inverse = false) {
+
+        auto &&num_qubits = this->getNumQubits();
+        std::size_t two2N =
+            std::exp2(num_qubits - wires.size() - controlled_wires.size());
+        std::size_t dim = std::exp2(wires.size());
+        KokkosVector matrix_trans("matrix_trans", matrix.size());
+
+        if (inverse) {
+            Kokkos::MDRangePolicy<DoubleLoopRank> policy_2d({0, 0}, {dim, dim});
+            Kokkos::parallel_for(
+                policy_2d,
+                KOKKOS_LAMBDA(const std::size_t i, const std::size_t j) {
```

Comment:
We don't need to specify `const` for `size_t`! Could you ensure that this is fixes everywhere?

Reply:
Why don't we need to specify `const` for `size_t`?

Reply:
I think our [cpp-dev-guide](https://github.com/PennyLaneAI/guidance-docs/blob/master/development/cpp-dev-guidelines.md) doesn't specify using `const` for primitive data type for parameter passed by values, but some guides like google [C++ guide](https://google.github.io/styleguide/cppguide.html#Use_of_const) suggests not? I am happy to take it all out to keep it consistent?

Reply:
I have taken them out in the LK `StateVectorKokkos.hpp` `BasicGeneratorFunctors.hpp` `MatrixGateFunctors.hpp` `BasicGateFunctors.hpp`

Reply:
Thanks @josephleekl! Yeah, adding `const` to PDT doesn't have any particular benefits for modern compilers. They are treated as SSA values w/ and w/o `const` anyway.